### ISSUE TITLE
Functions in TokenPropertyFunctions.java to have mapName as optional parameter

### ISSUE THREAD

**Issue by Merudo** [RPTools/maptool#545]
**Is your feature request related to a problem? Please describe.**
It is a pain to change properties of tokens on other maps. Instead of having to change the map and possibly create flickering, it would be better to add the mapName as additional optional argument.

**Describe the solution you'd like**
I would like the following functions in TokenImage.java to take a mapName parameter so they could work on tokens in other maps:

        getPropertyNames
        getPropertyNamesRaw
        hasProperty
        isNPC
        isPC
        setPC
        setNPC
        getLayer
        setLayer
        getSize
        setSize
        resetSize
        getOwners
        isOwnedByAll
        isOwner
        resetProperty
        getProperty
        setProperty
        isPropertyEmpty
        getPropertyDefault
        sendToBack
        bringToFront
        setPropertyType
        getPropertyType
        getRawProperty
        getTokenFacing
        setTokenFacing
        removeTokenFacing
        getTokenRotation
        getMatchingProperties
        isSnapToGrid
        setOwner
        setOwnedByAll
        getTokenNativeWidth
        getTokenNativeHeight
        getTokenWidth
        getTokenHeight
        setTokenWidth
        setTokenHeight
        getTokenShape
        setTokenShape
        getGMNotes
        setGMNotes
        getNotes
        setNotes
        setTokenSnapToGrid

**Comment by Azhrei** [comment]
I presume the fix for propagating tokens to other clients was solved by adding putToken(), but there are performance ramifications to adding it everywhere, which is why it hadn't already been done.

Every call to putToken() causes the entire token to be sent (which includes all properties and macros), and if you call putToken() twice, the token is sent twice.  There is no "queue" where attempts to send the token multiple times are collapsed into sending just the last one.

I don't think it's a good idea to add putToken() to all of these — imagine a macro that makes changes to a lot of properties all at once!

However, maybe there _should_ be a function that could push named tokens to clients under macro control?  This new `refreshClientTokens()` macro (or whatever it's called) could take a list of named tokens and call putToken() on each one.  Note that putToken() isn't really the right function to call — it should really call a lower-level function that just doesn't the client update, but such a function doesn't exist.  The entire putToken() call graph needs to be refactored; that could be a fair amount of work.  I looked at it some time ago but I don't remember the details.

**Comment by JamzTheMan** [comment]
I tried to look at that a little bit ago but didn't have time for that rabbit whole (and agree).  I suspect setState has putToken somewhere? My macros are horribly performing due to the token copies.

I thought maybe a macro to deferPutToken of some sort? Similar to your idea but instead call that this function at the start of your macro chain and set a boolean, put token is then only done once and only after all macros are done. (How ever that magic could be done, not sure if possible)

**Comment by Phergus** [comment]
Of course a lot of the set*() functions already had both a ServerCommand.putToken() and a Zone.putToken() call but not all.  Some had one and some had the other.  Which implies that some of the macro functions probably weren't propagating the changes out to the clients.

Note that the ServerCommand.putToken() does it's own call to Zone.putToken() so having both in the same macro is probably a bad thing.  But, the ServerCommand.putToken() method is calling Zone.putToken() on the current zone which is going to be a problem since the goal of this change is to update a token on a zone (i.e. map) that isn't the current one.

**Comment by Merudo** [comment]
Phergus: thanks for the observation that ServerCommand.putToken() does its own Zone.putToken(). I suggest all duplicates of Zone.putToken() be removed.

By the way, ServerCommand.putToken() takes zoneGUID as argument, so it can update a token on any zone.

**Comment by Merudo** [comment]
Also, is the performance issue with Zone.putToken(), or is it with ServerCommand.putToken() specifically?

My understanding is that Zone.putToken() updates the token locally, while ServerCommand.putToken() updates it both locally and on other clients.

**Comment by Merudo** [comment]
My problem with the lack of update is that if another client interacts with the token right away, the changes get canceled.

As for the general debate on the performance loss of ServerCommand.putToken(), I can see a few solutions:

1. Have some functions call putToken(), while some don't. It's not clear to me which functions should not call putToken(). In any case there should be some comments in the code justifying the lack of update. And as Azhrei said, a function refreshClientTokens() to push the update manually would be great.

2. Have a function, say, DisableUpdate(), disables all updates on the token until enabled back with EnableUpdate(). While the updates are disabled, refreshClientTokens() could be used to force the updates.

3. As JamzTheMan said, have a function deferPutToken() defers sending the updates until the macro is executed. I think it might be not as hard as anticipated. We could have deferPutToken() call EventQueue.invokeLater to defer putToken(). The code for execLink() could probably be reused.

4. Add a boolean argument to the macro functions, to decide if the changes should be pushed to clients or not.

I think option #3 with deferPutToken() option has a lot of merit. Basically, deferPutToken() would set a boolean variable to 1, which would disable ServerCommand.putToken(). Additionally, deferPutToken() would add an event with EventQueue.invokeLater to update the tokens.

The big question I have is, where should that boolean variable be located?

**Comment by Azhrei** [comment]
This is going to be a tough problem to solve with the current design re: how tokens are updated.  We’ve been down this rabbit hole before and I’d prefer to not change anything unless a comprehensive solution is presented, rather than a bandaid.

It may be that this should be tabled until the token split is achieved, since that will make token updates cleaner in the first place, and will likely necessitate some cleanup in this area in any case.

**Comment by Merudo** [comment]
That's Azhrei, but what should I do to my commit in the meanwhile? Should I just turn it back the way it was, with some functions set to not update clients?

**Comment by JamzTheMan** [comment]
I'm on the fence for this.  On one hand, It makes sense to hold off and properly design this with a split and differential updates (or what ever) and maybe a queue system (like rabbit mq or kafka) BUT....

I wouldn't also mind a band-aid either. If it greatly improved the clobbering of tokens and sped up state changes, It would help users over the next 6-12 months (AFAIK token split is NOT being worked on currently and such as, it could be over a year before we see anything on that front)

My 2 cents

**Comment by Azhrei** [comment]
And this is why I think putting it into the hands of the macro authors via a `refreshClients()` or similar macro function is the way to go.  It's a short-term bandaid.  In the future, we simply turn the function into a no-op and existing macros will continue to work fine.  (Or maybe it gets defined as a subset of the functionality it would need right now.)

In terms of the two putTokens(), with one calling the other, I guess a summary of what is done in each one would be necessary the best way to write the bandaid.  For example, is a json list of token IDs being passed to it?  Or is the queue of tokens kept internally?  Let's discuss it and see what comes out of the discussion...

**Comment by Phergus** [comment]
Clearly this problem changes significantly after the token split planned for 2.0 but right now it should be fairly straightforward to determine whether or not the ServerCommand.putToken() is sufficient and then remove the redundant Zone.putToken() calls if that turns out to be the case.

The ServerCommand.putToken() does use the Zone ID when it calls put Zone.putToken() so there shouldn't be an issue with updates on maps that aren't currently displayed.

The ServerCommand.putToken():
```
  public void putToken(GUID zoneGUID, Token token) {
    // Hack to generate zone event. All functions that update tokens call this method
    // after changing the token. But they don't tell the zone about it so classes
    // waiting for the zone change event don't get it.
    MapTool.getCampaign().getZone(zoneGUID).putToken(token);
    makeServerCall(COMMAND.putToken, zoneGUID, token);
  }
```
The set*() functions should be tested against 1.5.3 to see if they are updating client tokens or not and then be tested again with the Zone.putToken() calls removed and only the ServerCommand.putToken() used.

**Comment by JamzTheMan** [comment]
Agreed, let's do some analysis and see what the easiest and best fix is. I'm ok with a new function as well if we can remove some of the redundancy.

**Comment by Merudo** [comment]
I did a bit more research into Zone.putToken().

As Phergus said, the first thing ServerCommand.putToken() does is [call Zone.putToken()](https://github.com/RPTools/maptool/blob/9191a135d11d28ebd7e2e8cf122cf08b59ceb139/src/main/java/net/rptools/maptool/client/ServerCommandClientImpl.java#L122). So, there is really no justification for the duplication.

Now, sometimes you may want to update a token locally and but don't want to update remote clients for performance issues. In those cases, you probably don't need to use [Zone.putToken()](https://github.com/RPTools/maptool/blob/9191a135d11d28ebd7e2e8cf122cf08b59ceb139/src/main/java/net/rptools/maptool/model/Zone.java#L1129) either. Since the object "token" is a reference every change to the token will actually update correctly without using putToken(). The only justification for putToken()  is to trigger FireModelChangeEvent(), which could be fired without using putToken() anyway...

Avoiding putToken() seems to be a good idea, since each time it is run, the whole token is deleted from the zone, then added back. The list of token is then sorted. I can see those operations being very costly.

**Comment by Merudo** [comment]
> I tried to look at that a little bit ago but didn't have time for that rabbit whole (and agree). I suspect setState has putToken somewhere? My macros are horribly performing due to the token copies.

[It does](https://github.com/RPTools/maptool/blob/9191a135d11d28ebd7e2e8cf122cf08b59ceb139/src/main/java/net/rptools/maptool/client/functions/TokenStateFunction.java#L180). The result is a call to [ServerCommand.putToken()](https://github.com/RPTools/maptool/blob/9191a135d11d28ebd7e2e8cf122cf08b59ceb139/src/main/java/net/rptools/maptool/client/MapToolVariableResolver.java#L450).

**Comment by Phergus** [comment]
> By the way, what is the rationale for updating changes to other clients by sending them the whole token? Why not send them the instruction to update individual token properties? 

That decision was made 14+ years ago.  If you go back to the beginnings in the forums you might find some discussion of it.  Originally MapTool didn't have a macro language and changes made to tokens were all done through the Edit Token dialog or right-click menu.  At the time, tokens had little content and the single image (no portrait or handout) were were mostly low-res only a couple KB in size.  There have been many discussions since then about the whole token transfer issue as well as avoiding concurrent modifications but those discussions haven't turned into action yet so here we are.

**Comment by Merudo** [comment]
I'd love to read the previous discussion on this issue, but I can't seem to find much.

**Comment by JamzTheMan** [comment]
IIRC states are not set on tokens, but rather in their own map, correct? Why is putToken (or the serverCommand) called when setState is called? Anyone know?

**Comment by Azhrei** [comment]
Um, because there is no provision for sending anything smaller than the Token?  That’s the whole issue with tokens — They’ve gotten larger and larger and yet we continue to send the whole thing.

States would be in of the first beneficiaries of the token split since their

**Comment by Phergus** [comment]
> IIRC states are not set on tokens, but rather in their own map, correct? Why is putToken (or the serverCommand) called when setState is called? Anyone know?

The states are saved with the tokens but I guess that doesn't necessarily mean that they aren't managed at the map level.  But now I'm afraid to go look.  lol

**Comment by Phergus** [comment]
Closed by accident.

**Comment by Merudo** [comment]
Maybe I'm missing something? It doesn't seem too hard to add methods to send data smaller than the Token? We just need to add them to ServerCommandClientImp.java, ServerCommand.java, and ServerMethodHandlerjava.

**Comment by JamzTheMan** [comment]
@phergus and @azhrei I don't believe states are part of the token class so why putToken? I believe there is just a hashmap or such that stores token id/state ?

**Comment by Merudo** [comment]
[States are part of the token class](https://github.com/RPTools/maptool/blob/9191a135d11d28ebd7e2e8cf122cf08b59ceb139/src/main/java/net/rptools/maptool/model/Token.java#L216) I'm afraid.

**Comment by JamzTheMan** [comment]
Gah, ok, thanks. Was going off memory and. I remember going down a rabbit hole as states can be changed via several different places in the code.

**Comment by Phergus** [comment]
> It doesn't seem too hard to add methods to send data smaller than the Token? We just need to add them to ServerCommandClientImp.java, ServerCommand.java, and ServerMethodHandlerjava.

@Merudo  I believe you are correct.  Note this would not be done under this issue but would be done under a separate issue.  First step would be to describe the proposed new Server Commands.  Obvious candidates would be ones to handle setting properties and states as these are actions taken constantly in framework macros.

My thought is that new server commands should probably be pushed to 1.5.5.  The changes to macro functions for supporting a mapname parameter can be part of 1.5.4.   **If** testing shows that the Zone.putToken() calls are redundant those calls could be removed for the functions that are being changed under this issue.  Broader removal of Zone.putToken() calls from other functions should also get pushed to 1.5.5 once 1.5.4 has had further testing to have some confidence that the changes haven't broken anything.

**Comment by JamzTheMan** [comment]
Perhaps we create a new issue and continue discussions? This has gotten far off rails for this issue (although this issue could depend/refer to this new issue)

**Comment by Phergus** [comment]
Agreed.

**Comment by Phergus** [comment]
All functions which modify tokens and are noted above should be reviewed to determine if they are making the redundant Zone.putToken() call (which only updates local client) and if so it should be removed from each.  Checking of course that each one does use the ServerCommand.putToken() call to ensure that changes propagate to clients.

**Comment by Merudo** [comment]
@Phergus, @Azhrei, and @JamzTheMan, I made a proof of concept commit where the function setProperty() is changed to update clients by only sending the modified property #556. We could discuss it further in #555.

**Comment by Phergus** [comment]
Note that all macro function changes need to be documented on the Wiki.

**Comment by Phergus** [comment]
What is the current status of this?  Have all the listed macro functions been modified to take a map name as a parameter?

**Comment by Merudo** [comment]
Yes! But the documentation needs to be updated.

**Comment by Phergus** [comment]
Documentation:
 - [ ] [bringToFront](http://www.lmwcs.com/rptools/wiki/bringToFront)
 - [ ] [sendToBack](http://www.lmwcs.com/rptools/wiki/sendToBack)
 - [ ] [getGMNotes](http://www.lmwcs.com/rptools/wiki/getGMNotes)
 - [ ] [getLayer](http://www.lmwcs.com/rptools/wiki/getLayer)
 - [ ] [getMatchingProperties](http://www.lmwcs.com/rptools/wiki/getMatchingProperties)
 - [ ] [getNotes](http://www.lmwcs.com/rptools/wiki/getNotes)
 - [ ] [getOwners](http://www.lmwcs.com/rptools/wiki/getOwners)
 - [ ] [getProperty](http://www.lmwcs.com/rptools/wiki/getProperty)
 - [ ] [getPropertyDefault](http://www.lmwcs.com/rptools/wiki/getPropertyDefault)
 - [ ] [getPropertyNames](http://www.lmwcs.com/rptools/wiki/getPropertyNames)
 - [ ] [getPropertyNamesRaw](http://www.lmwcs.com/rptools/wiki/getPropertyNamesRaw)
 - [ ] [getPropertyType](http://www.lmwcs.com/rptools/wiki/getPropertyType)
 - [ ] [getRawProperty](http://www.lmwcs.com/rptools/wiki/getRawProperty)
 - [ ] [getSize](http://www.lmwcs.com/rptools/wiki/getSize)
 - [ ] [getTokenFacing](http://www.lmwcs.com/rptools/wiki/getTokenFacing)
 - [ ] [getTokenHeight](http://www.lmwcs.com/rptools/wiki/getTokenHeight)
 - [ ] [getTokenNativeHeight](http://www.lmwcs.com/rptools/wiki/getTokenNativeHeight)
 - [ ] [getTokenNativeWidth](http://www.lmwcs.com/rptools/wiki/getTokenNativeWidth)
 - [ ] [getTokenRotation](http://www.lmwcs.com/rptools/wiki/getTokenRotation)
 - [ ] [getTokenShape](http://www.lmwcs.com/rptools/wiki/getTokenShape)
 - [ ] [getTokenWidth](http://www.lmwcs.com/rptools/wiki/getTokenWidth)
 - [ ] [hasProperty](http://www.lmwcs.com/rptools/wiki/hasProperty)
 - [ ] [isNPC](http://www.lmwcs.com/rptools/wiki/isNPC)
 - [ ] [isOwnedByAll](http://www.lmwcs.com/rptools/wiki/isOwnedByAll)
 - [ ] [isOwner](http://www.lmwcs.com/rptools/wiki/isOwner)
 - [ ] [isPC](http://www.lmwcs.com/rptools/wiki/isPC)
 - [ ] [isPropertyEmpty](http://www.lmwcs.com/rptools/wiki/isPropertyEmpty)
 - [ ] [isSnapToGrid](http://www.lmwcs.com/rptools/wiki/isSnapToGrid)
 - [ ] [removeTokenFacing](http://www.lmwcs.com/rptools/wiki/removeTokenFacing)
 - [ ] [resetProperty](http://www.lmwcs.com/rptools/wiki/resetProperty)
 - [ ] [resetSize](http://www.lmwcs.com/rptools/wiki/resetSize)
 - [ ] [setGMNotes](http://www.lmwcs.com/rptools/wiki/setGMNotes)
 - [ ] [setLayer](http://www.lmwcs.com/rptools/wiki/setLayer)
 - [ ] [setNPC](http://www.lmwcs.com/rptools/wiki/setNPC)
 - [ ] [setNotes](http://www.lmwcs.com/rptools/wiki/setNotes)
 - [ ] [setOwnedByAll](http://www.lmwcs.com/rptools/wiki/setOwnedByAll)
 - [ ] [setOwner](http://www.lmwcs.com/rptools/wiki/setOwner)
 - [ ] [setPC](http://www.lmwcs.com/rptools/wiki/setPC)
 - [ ] [setProperty](http://www.lmwcs.com/rptools/wiki/setProperty)
 - [ ] [setPropertyType](http://www.lmwcs.com/rptools/wiki/setPropertyType)
 - [ ] [setSize](http://www.lmwcs.com/rptools/wiki/setSize)
 - [ ] [setTokenFacing](http://www.lmwcs.com/rptools/wiki/setTokenFacing)
 - [ ] [setTokenHeight](http://www.lmwcs.com/rptools/wiki/setTokenHeight)
 - [ ] [setTokenShape](http://www.lmwcs.com/rptools/wiki/setTokenShape)
 - [ ] [setTokenSnapToGrid](http://www.lmwcs.com/rptools/wiki/setTokenSnapToGrid)
 - [ ] [setTokenWidth](http://www.lmwcs.com/rptools/wiki/setTokenWidth)

**Comment by Phergus** [comment]
Code changes and documentation complete.

**Comment by Phergus** [comment]
@Merudo  Did you decide to not add ID & Map to getPropertyDefault()?

**Comment by Phergus** [comment]
All functions tested with optional ID and map name parameters successfully with the exception of getPropertyDefault().

**Comment by Merudo** [comment]
Yes I did not add ID & mapName to getPropertyDefault().

I thought it would be confusing because the second argument to getPropertyDefault() is actually "propType" (token type defined in Campaign Properties), not a token.
